lib/compat/wordpress-6.2/get-global-styles-and-settings.php

```diff
@@ -45,3 +45,108 @@ function wp_theme_clean_theme_json_cached_data() {
 		WP_Theme_JSON_Resolver_Gutenberg::clean_cached_data();
 	}
 }
+
+/**
+ * Returns the stylesheet resulting of merging core, theme, and user data.
+ *
+ * @param array $types Types of styles to load. Optional.
+ *                     It accepts 'variables', 'styles', 'presets' as values.
+ *                     If empty, it'll load all for themes with theme.json support
+ *                     and only [ 'variables', 'presets' ] for themes without theme.json support.
+ *
+ * @return string Stylesheet.
+ */
+function gutenberg_get_global_stylesheet( $types = array() ) {
+	// Return cached value if it can be used and exists.
+	// It's cached by theme to make sure that theme switching clears the cache.
+	$can_use_cached = (
+		( empty( $types ) ) &&
+		( ! defined( 'WP_DEBUG' ) || ! WP_DEBUG ) &&
+		( ! defined( 'SCRIPT_DEBUG' ) || ! SCRIPT_DEBUG ) &&
```

Comment:
Those will be truthy in our phpunit setup, so a test will not be able to use the cached data in the current state in a test. We could, of course, add a filter to circumvent that 😅

Reply:
Let's remove this, as it hides caching behaviour changes on local dev, with it not great for unit testing and local profiling.

Reply:
If we don't have a mechanism to invalidate the cache in development mode, what would you recommend to provide a good developer experience (changes to `theme.json` are immediately applied)?

Reply:
> We could, of course, add a filter to circumvent that 😅

Yeah, I discussed this more deeply with @oandregal and probably will give that a try, so we can keep the cache off by default in some scenarios (e.g. during theme development). There might be also legitimate scenarios where plugins want to disable this cache as well (e.g. if they want to change the styles in _real time_ using some of the Global Styles filters), so they could benefit from this filter as well.

Reply:
In https://github.com/WordPress/gutenberg/pull/45679/commits/a845e8538d1c04c497f4c4bdf6ebf7e9c2359e61 I simplified the conditions that allows the cache to be used by removing the `! REST_REQUEST` and `! is_admin()` checks, so from now on we should be caching more often.

I kept the `! WP_DEBUG` and `! SCRIPT_DEBUG` checks since theme developers presumably have these constants turned on when developing a theme and we don't want the cache to interfere with their work.

I also kept the `empty( $types )` check, since otherwise we would need to cache the stylesheet for every possible value in `$types`.

Furthermore, there is now a filter that can override the default cache behavior when needed such as in the unit tests.

Reply:
(I want to document some of the rationale for the filter)

The use case the filter addresses is allowing consumers to control cache based on external conditions. Sometimes, it's not enough to clear the cache upon all the events this is hooked into.

Initially, my thinking was that Miguel and I discussed that this could be enough (asking consumers to clear the cache under those conditions):

```php
// In some 3rd party code, such as a plugin:
if ( $some_conditions_are_met ) {
	gutenberg_get_global_stylesheet_clean_cache();
	// So it gets recalculated based on
	// the results of the filters for theme.json data.
}

// ...

// In core code:
gutenberg_get_global_stylesheet(); // will repopulate the cache
```

However, this code is vulnerable to race conditions.

A simple example would be when the object cache is implemented as a persistent cache across request that is shared among many front-end servers. Other examples exist, but this is easy to visualize. A reproducible race condition would be:

- server A cleans the cache (in the plugin code)
- server B sets the cache (in the core code)
- server A goes to find the value (in the core code) and it comes from the cache, so it's not recalculated as expected

By allowing the plugin code to modify the core behaviour (by means of the filter):

```php
function gutenberg_get_global_stylesheet( $types = array() ) {
	$can_use_cached = apply_filters( 'wp_get_global_stylesheet_can_use_cache', $default );
	// ...
```

the cache will never be used if the consumer does not want to, avoiding race conditions.